### PR TITLE
BUG: regression for array([pandas.DataFrame()])

### PR DESCRIPTION
**Pull request by mattip**
Fixes #13399

Add failing test. 
Fix by not converting 0-len object to ndarray, which is compatible with behaviour before #13399

### CODE REVIEW

**Comment by mattip** [comment]
Fixes #13659

**Comment by mattip** [comment]
xref pandas-dev/pandas#26546

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -477,32 +472,40 @@ setArrayFromSequence(PyArrayObject *a, PyObject *s,
...
-        Py_DECREF(s);
-        return 0;
-    }
-    else {
-        Py_DECREF(Py_NotImplemented);
+            Py_DECREF(s);
+            return 0;
+        }
+        else {
+            Py_DECREF(Py_NotImplemented);
```
Comment (seberg) [inline numpy/core/src/multiarray/ctors.c]: Shouldn't this be an unconditional `Py_DECREF(tmp)`?

File: numpy/core/tests/test_regression.py [hunk 2]
```diff
@@ -2454,3 +2454,43 @@ class T(object):
...
+
+            def __len__(self):
+                return 0
+
+
+        t = T()
+        #gh-13659, would raise in broadcasting [x=t for x in result]
+        np.array([t])
+
+    def test_2d__array__shape(self):
```
Comment (seberg) [inline numpy/core/tests/test_regression.py]: Seems like this whole code block is just duplicated?

**Review by seberg — COMMENTED** [review]
Seems OK to me as a reversion. I am not sure about the DECREF, I was wondering if it should have a comment that this is a bit odd, but then if its a 0 length sequence, the assignment *should* be empty in any case (of course Eric will create some malicious object which would bring it down in flames).

Fixes the regression hitting Pandas. But does not fix the real (never noticed) but that this gives the incorrect output shape to begin with.

File: numpy/core/tests/test_regression.py (hunk 2 above)
Comment (mattip) [reply numpy/core/tests/test_regression.py]: weird. Fixing

File: numpy/core/src/multiarray/ctors.c (hunk 1 above)
Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: fixing

File: numpy/core/src/multiarray/ctors.c [hunk 3]
```diff
@@ -477,32 +471,39 @@ setArrayFromSequence(PyArrayObject *a, PyObject *s,
...
+    if (slen > 0) {
+        /* gh-13659: objects with 0 len can still have ndim > ndim(dst) */
+        PyObject *tmp = _array_from_array_like(s, /*dtype*/NULL, /*writeable*/0,
+                                               /*context*/NULL);
+        if (tmp == NULL) {
             goto fail;
         }
+        else if (tmp != Py_NotImplemented) {
+            if (PyArray_CopyInto(dst, (PyArrayObject *)tmp) < 0) {
+                goto fail;
```
Comment (seberg) [inline numpy/core/src/multiarray/ctors.c]:
```suggestion
                Py_DECREF(tmp);
                goto fail;
```
Something is still off here with the reference counting. Aso the return 0 below needs it. Which probably means the whether or not the else clause is there did not matter much. Or we make `tmp = NULL;` one scope up to make the fail paths nicer.

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: refactored

Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Thanks, looks right now. Will merge as soon as tests finish (or squash merge).

File: numpy/core/tests/test_regression.py [hunk 4]
```diff
@@ -2454,3 +2454,23 @@ class T(object):
...
+            def __getitem__(self, idx):
+                raise AssertionError("__getitem__ was called")
+
+            def __len__(self):
+                return 0
+
+
+        t = T()
+        #gh-13659, would raise in broadcasting [x=t for x in result]
+        np.array([t])
```
Comment (eric-wieser) [inline numpy/core/tests/test_regression.py]: Can we add a test here for the shape?

Comment (mattip) [reply numpy/core/tests/test_regression.py]: xref #13671

**Pull request merged**